File: src/02.lua
require 'src.common.common'

local actions = {
  forward = function(submarine, value)
    submarine.hpos = submarine.hpos + value
    submarine.depth_2 = submarine.depth_2 + submarine.aim * value
  end,
  up = function(submarine, value)
    submarine.depth_1 = submarine.depth_1 - value
    submarine.aim = submarine.aim - value
  end,
  down = function(submarine, value)
    submarine.depth_1 = submarine.depth_1 + value
    submarine.aim = submarine.aim + value
  end,
}

function solve(commands)
  local submarine = {hpos = 0, depth_1 = 0, depth_2 = 0, aim = 0}
  for command in all(commands) do
    actions[command.direction](submarine, command.value)
  end
  return submarine.hpos * submarine.depth_1, submarine.hpos * submarine.depth_2
end

function parse(line)
  local match_it = line:gmatch('(%w+) (%w+)')
  local d, v = match_it()
  return {direction = d, value = tonumber(v)}
end

local commands = getParsedLines(parse)
local result_1, result_2 = solve(commands)
print(result_1)
print(result_2)

File: src/02.rs
#[path = "common/common.rs"]
mod common;

use common::get_parsed_lines;

fn solve(commands: &Vec<Command>) -> (i32, i32) {
    let (mut hpos, mut depth_1, mut depth_2, mut aim) = (0, 0, 0, 0);
    for command in commands {
        match command.direction.as_str() {
            "forward" => {
                hpos += command.value;
                depth_2 += aim * command.value;
            }
            "up" => {
                depth_1 -= command.value;
                aim -= command.value;
            }
            "down" => {
                depth_1 += command.value;
                aim += command.value;
            }
            _ => panic!(),
        }
    }
    (hpos * depth_1, hpos * depth_2)
}

struct Command {
    direction: String,
    value: i32,
}

fn parse(line: String) -> Command {
    let chunks: Vec<_> = line.split_whitespace().collect();
    let direction = chunks[0].to_string();
    let value = chunks[1].parse().unwrap();

    Command{ direction, value }
}

fn main() {
    let commands = get_parsed_lines(parse);
    let (result_1, result_2) = solve(&commands);
    println!("{result_1}\n{result_2}");
}
